### ISSUE TITLE
Provide a way to prevent cops from being disabled

### ISSUE THREAD

**Issue by rafaelfranca** [rubocop/rubocop#14598]
In my applications I want to prevent `rubocop:disable` and `rubocop:todo` to be used.

See https://github.com/Shopify/ruby-style-guide/pull/756 for the implementation.

But the problem I'm facing now is that you can disable the cop that prevent the disabling of cops.

I see that some cops [can't be disabled like `Lint/Syntax`](https://github.com/rubocop/rubocop/pull/12760).

Is there a way to provide that API to cops?

If not, would you willing to accept the cop I'm writing in my library into the Lint department? That way I can do the same as the `Lint/Syntax`.

**Comment by Earlopain** [comment]
Personally I think such a cop would be ok for rubocop itself (if disabled by default).

But I am not so sure about ensuring that it can't be disabled. Unsafe cops sometimes explicitly call out code where there may be false positives, what would you do in such a case? You may be able to rewrite code in a different way so as to not trigger rubocop but I don't think that would be a very nice developer experience. Perhaps you can't rewrite differently at all.

I see that you implemented `AllowedCops`, would it not be better to allow it to be disabled case-by-case?

**Comment by jeromedalbert** [comment]
There is already a cop called [`Style/DisableCopsWithinSourceCodeDirective`](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb) that was added in https://github.com/rubocop/rubocop/pull/7384. It supports both `rubocop:disable` and `rubocop:todo`.

I had the cop configured like so in a past project's `.rubocop.yml`:

```yaml
Style/DisableCopsWithinSourceCodeDirective:
  # This is enabled to make sure that every RuboCop error gets fixed and not
  # quickly disabled with a comment.
  #
  # - DO *NOT* disable this setting.
  # - DO *NOT* work around this rule by editing the .rubocop_todo.yml, as this
  # file is only for old legacy code.
  Enabled: true
```

This may not scale well as this relies on trust or heavy monitoring of any changes to the `.rubocop.yml` and `.rubocop_todo.yml` files.

So I think it would be nice to prevent such a cop from being disabled, but I think it should only be for the `DisableCopsWithinSourceCodeDirective` cop. Because I fear that a generic option to prevent cops from being disabled could be abused by owners of rubocop collections (for example `standardrb/standard`) to mark all their cops as non-disableable, which would stray from the original intent of such an option, and be inflexible and tedious to work around if extra customization is needed.

**Comment by Earlopain** [comment]
Oh, great catch. I did not realize it already existed. I'm still rather negative on it not being able to be disabled though.

I wonder if that behavior could be controlled via opt-in through the cop config, that would be nice I think.

**Comment by rafaelfranca** [comment]
> There is already a cop called [Style/DisableCopsWithinSourceCodeDirective](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb) that was added in https://github.com/rubocop/rubocop/pull/7384

Oh. Thank you for pointing that. It seems that cop does eveyrthing I want.

> Unsafe cops sometimes explicitly call out code where there may be false positives, what would you do in such a case?

That is why I added the `AllowedCops` option. You can add cops in that list that you expect to sometimes disable the cop with a directive. The `html_safe` related cop is one that come to mind.

But what I want is to not allow disabling the `Style/DisableCopsWithinSourceCodeDirective` cop.

> I wonder if that behavior could be controlled via opt-in through the cop config, that would be nice I think.

How could we do that? Looking at the implementation, the only way to disallow disabling cops using directives is with the hardcoded list in `exclude_lint_department_cops`

**Comment by corsonknowles** [comment]
We'd be very interested in using [Style/DisableCopsWithinSourceCodeDirective](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb) with a `RequiredCops` shortlist instead of allowing 1000 rules to protect 5.
